src/OpenFOAM/matrices/lduMatrix/lduMatrix/lduMatrixATmul.C

```diff
@@ -75,64 +74,130 @@ struct matrixMultiplyFunctor
     __device__
     scalar operator()(const label& id) const
     {
-        scalar tmpSum[2*nUnroll] = {};
-        scalar nExtra = 0;
+        #if __CUDA_ARCH__ >= 350
```

Comment:
This introduces a lot of duplicated code which is a really really bad idea, especially here where the logic of what is going on is not trivial.